File: mf2/java/src/test/java/com/box/l10n/mojito/mf2/OfficialBridge.java
package com.box.l10n.mojito.mf2;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/** Test-only JSON-lines adapter for the shared upstream assertion checker. */
public final class OfficialBridge {
    private OfficialBridge() {}
    @SuppressWarnings("unchecked")
    public static void main(String[] args) throws Exception {
        var input = new BufferedReader(new InputStreamReader(System.in, StandardCharsets.UTF_8));
        for (String line; (line = input.readLine()) != null;) {
            Map<String, Object> request = (Map<String, Object>) JsonParser.parse(line);
            var parsed = Mf2Parser.parseToModel((String) request.get("source"));
            Object response;
            if (parsed.hasDiagnostics()) {
                response = Map.of("diagnostics", parsed.diagnostics().stream().map(Mf2ParseDiagnostic::code).toList(), "value", "", "errors", List.of(), "parts", List.of());
            } else {
                var base = "platform".equals(request.get("registry")) ? Mf2FunctionRegistry.defaults() : Mf2FunctionRegistry.portable();
                var options = Mf2FormatOptions.builder().locale((String) request.getOrDefault("locale", "en"))
                        .bidiIsolation(Mf2BidiIsolation.fromName((String) request.getOrDefault("bidiIsolation", "none")))
                        .functions(ConformanceFunctionRegistry.withTestFunctions(base)).build();
                var arguments = (Map<String, Object>) request.getOrDefault("arguments", Map.of());
                try {
                    var result = parsed.model().format(arguments, options);
                    var parts = parsed.model().formatToParts(arguments, options);
                    response = Map.of("diagnostics", List.of(), "value", result.value(), "errors", result.errors().stream().map(Mf2Exception::code).toList(),
                            "parts", FormattedPartJson.toMaps(parts.parts()), "partsErrors", parts.errors().stream().map(Mf2Exception::code).toList());
                } catch (Mf2Exception error) {
                    response = Map.of("diagnostics", List.of(), "value", "", "errors", List.of(error.code()), "parts", List.of(), "partsErrors", List.of(error.code()));
                }
            }
            System.out.println(json(response));
        }
    }
    static String json(Object value) {
        if (value == null) return "null";
        if (value instanceof String text) {
            StringBuilder output = new StringBuilder("\"");
            for (int index = 0; index < text.length(); index++) {
                char ch = text.charAt(index);
                switch (ch) {
                    case '"' -> output.append("\\\"");
                    case '\\' -> output.append("\\\\");
                    case '\n' -> output.append("\\n");
                    case '\r' -> output.append("\\r");
                    case '\t' -> output.append("\\t");
                    default -> { if (ch < 0x20) output.append(String.format("\\u%04x", (int) ch)); else output.append(ch); }
                }
            }
            return output.append('"').toString();
        }
        if (value instanceof Map<?, ?> map) return map.entrySet().stream().map(entry -> json(entry.getKey().toString()) + ":" + json(entry.getValue())).collect(Collectors.joining(",", "{", "}"));
        if (value instanceof List<?> list) return list.stream().map(OfficialBridge::json).collect(Collectors.joining(",", "[", "]"));
        return value.toString();
    }
}

File: mf2/php/tests/official_bridge.php
<?php

declare(strict_types=1);

use Mojito\MessageFormat2\FunctionRegistry;
use Mojito\MessageFormat2\IntlFunctions;
use function Mojito\MessageFormat2\parse_to_model;
use function Mojito\MessageFormat2\format_message;
use function Mojito\MessageFormat2\format_message_to_parts;
use function Mojito\MessageFormat2\Internal\error_code;

define('MF2_OFFICIAL_BRIDGE', true);
require_once __DIR__ . '/unicode_tests.php';

while (($line = fgets(STDIN)) !== false) {
    $request = json_decode($line, true, 512, JSON_THROW_ON_ERROR);
    $parsed = parse_to_model($request['source']);
    $response = ['diagnostics' => array_column($parsed['diagnostics'], 'code'), 'value' => '', 'errors' => [], 'parts' => []];
    if (!$parsed['hasDiagnostics']) {
        $base = ($request['registry'] ?? 'portable') === 'platform' ? IntlFunctions::registry() : FunctionRegistry::portable();
        $registry = $base->withFunction('test:function', 'official_test_function')->withFunction('test:select', 'official_test_select_resolver')->withFunction('test:format', 'official_test_format_resolver')
            ->withSelector('test:function', 'official_test_selector')->withSelector('test:select', 'official_test_selector')->withSelector('test:format', 'official_test_format_selector');
        $options = ['locale' => $request['locale'] ?? 'en', 'bidiIsolation' => $request['bidiIsolation'] ?? 'none', 'functions' => $registry];
        $result = format_message($parsed['model'], $request['arguments'] ?? [], $options);
        $parts = format_message_to_parts($parsed['model'], $request['arguments'] ?? [], $options);
        $response['value'] = $result['value'];
        $response['errors'] = array_map(error_code(...), $result['errors']);
        $response['parts'] = $parts['parts'];
        $response['partsErrors'] = array_map(error_code(...), $parts['errors']);
    }
    echo json_encode($response, JSON_THROW_ON_ERROR | JSON_UNESCAPED_UNICODE) . "\n";
}
